controllers/replica_set_controller.go

```diff
@@ -98,50 +94,35 @@ func init() {
 	zap.ReplaceGlobals(logger)
 }
 
-// Add creates a new MongoDB Controller and adds it to the Manager. The Manager will set fields on the Controller
-// and Start it when the Manager is Started.
-func Add(mgr manager.Manager) error {
-	return add(mgr, newReconciler(mgr, readVersionManifestFromDisk))
-}
-
 // ManifestProvider is a function which returns the VersionManifest which
 // contains the list of all available MongoDB versions
 type ManifestProvider func() (automationconfig.VersionManifest, error)
 
-func newReconciler(mgr manager.Manager, manifestProvider ManifestProvider) *ReplicaSetReconciler {
+func NewReconciler(mgr manager.Manager, manifestProvider ManifestProvider) *ReplicaSetReconciler {
 	mgrClient := mgr.GetClient()
 	secretWatcher := watch.New()
 
+	mp := manifestProvider
+	if mp == nil {
+		mp = readVersionManifestFromDisk
+	}
+
 	return &ReplicaSetReconciler{
 		client:           kubernetesClient.NewClient(mgrClient),
 		scheme:           mgr.GetScheme(),
-		manifestProvider: manifestProvider,
+		manifestProvider: mp,
 		log:              zap.S(),
 		secretWatcher:    &secretWatcher,
 	}
 }
 
-// add sets up a controller for the Reconciler on the manager. It will
-// also configure the necessary watches.
-func add(mgr manager.Manager, r *ReplicaSetReconciler) error {
-	// Create a new controller
-	c, err := controller.New("replicaset-controller", mgr, controller.Options{Reconciler: r})
-	if err != nil {
-		return err
-	}
-
-	// Watch for changes to primary resource MongoDB
-	err = c.Watch(&source.Kind{Type: &mdbv1.MongoDBCommunity{}}, &handler.EnqueueRequestForObject{}, predicates.OnlyOnSpecChange())
-	if err != nil {
-		return err
-	}
-
-	err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, r.secretWatcher)
-	if err != nil {
-		return err
-	}
-
-	return nil
+// SetupWithManager sets up the controller with the Manager and configures the necessary watches.
+func (r *ReplicaSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
```

Comment:
[q] is there a reason this is a method and not a function? Does this implement some interface that is required?

Not a big deal either way!

Reply:
I don't think it's required, but it is the way the scaffolding intended (and matches the Reconcile method). I don't think this particular method gets called anywhere outside of main.go (e.g. by the library), but my vote would be to keep it consistent with operator-sdk's expectations so that we don't diverge too far, just in case they introduce code expecting it to follow this format.